application/src/main/java/org/togetherjava/tjbot/features/cakeday/CakeDayListener.java

```diff
@@ -0,0 +1,84 @@
+package org.togetherjava.tjbot.features.cakeday;
+
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.entities.Member;
+import net.dv8tion.jda.api.entities.User;
+import net.dv8tion.jda.api.events.guild.member.GuildMemberRemoveEvent;
+import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
+import net.dv8tion.jda.api.hooks.ListenerAdapter;
+import org.jetbrains.annotations.NotNull;
+
+import org.togetherjava.tjbot.db.generated.tables.records.CakeDaysRecord;
+import org.togetherjava.tjbot.features.EventReceiver;
+
+import java.util.Optional;
+
+/**
+ * A listener class responsible for handling cake day related events.
+ */
+public class CakeDayListener extends ListenerAdapter implements EventReceiver {
+
+    private final CakeDayService cakeDayService;
+
+    /**
+     * Constructs a new CakeDayListener with the given {@link CakeDayService}.
+     *
+     * @param cakeDayService the {@link CakeDayService} to be used by this listener
+     */
+    public CakeDayListener(CakeDayService cakeDayService) {
+        this.cakeDayService = cakeDayService;
+    }
+
+    /**
+     * Handles the event of a message being received in a guild.
+     * <p>
+     * It caches the user's cake day and inserts the member's cake day into the database if not
+     * already present.
+     *
+     * @param event the {@link MessageReceivedEvent} representing the message received
+     */
+    @Override
+    public void onMessageReceived(@NotNull MessageReceivedEvent event) {
```

Comment:
```java
LocalDate currentDate = LocalDate.now();
                List<Member> anniversaries = guild
                        .loadMembers()
                        .get()
                        .stream()
                        .filter(member -> member.getTimeJoined().getDayOfMonth() == currentDate.getDayOfMonth() && member.getTimeJoined().getMonth() == currentDate.getMonth())
                        .toList();
```

Why can't we run this as a routine regularly?

Reply:
Wouldn't this include "0-day" anniversaries?

Reply:
You can update the filter to include `member.getTimeJoined().getYear() != currentDate.getYear()`

Reply:
> ```java
> LocalDate currentDate = LocalDate.now();
> List<Member> anniversaries = guild
>   .loadMembers()
>   .get()
>   .stream()
>   .filter(member -> member.getTimeJoined().getDayOfMonth() == currentDate.getDayOfMonth() && member.getTimeJoined().getMonth() == currentDate.getMonth())
>   .toList();
> ```
> 
> Why can't we run this as a routine regularly?

Wouldn't this be expensive to run once every day? It has to go through all members every time and there's been some hurdles with Discord's rate limiter, forcing us to use a database for this.

Reply:
That snippet will be expensive but we also know that JDA caches on startup. So use the cached members when doing the loop.

Reply:
This resolved?